use setBuffer instead of mutating buffer

AhmadMayo opened the pull request:
`useEdit` hook was mutating the `buffer` variable directly instead of using `useBuffer`. The ui was being updated because of the call to `setIsMSN`, which re-renders the `HexView` component.

I don't know whether or not this actually fixes a bug, but I feel that it - at least - prevents a potential bug from occurring.

P.S. thanks for the video, I learned a lot from you.

Commit:
use setBuffer instead of mutating buffer

Commit:
remove whitespaces inserted from formatting

AhmadMayo commented on src/hooks/use-edit.ts:
```diff
@@ -24,24 +25,27 @@ export const useEdit = ({
 
   useEffect(() => {
     setIsMSN(true);
-  }, [cursor, buffer]);
```
We can't reset the flag on each update, because we update the buffer when the user makes an edit.
Instead, `setIsMSN(true)` was called after each `bufferCommand` call at lines 63, 70, and 75.

francisrstokes replied:
We can put the buffer back into this dependency array now

AhmadMayo committed:
remove another whitespace weirdly inserted by formatting

AhmadMayo committed:
some more whitespaces 😶

AhmadMayo committed:
whitepaces

AhmadMayo committed:
more whitespaces

AhmadMayo commented:
I didn't see the end of the video when I made this, so sorry for this.
But I actually benchmarked it for a buffer of length 2**16, and it took 112 ms for a 1000 iterations, so I guess there's no performance issue?

Here's the test code

```javascript
function performanceTest(testFunction, iterations) {
  const sum = 0;
  const start = performance.now();
  for (let i = 0; i < iterations; i++) {
    testFunction();
  }
  
  return performance.now() - start;
}

numBytes = 2**16;
generateRandomByte = () => (Math.random() * 256) | 0;

bytes = Array.from({ length: numBytes }, generateRandomByte);
buffer = new Uint8Array(bytes);

function myFunc() {
  const newBuffer = buffer.slice();
  console.log(buffer.length);

  return newBuffer;
}

performanceTest(myFunc, 1000);
```

francisrstokes commented:
Hi Ahmed, thanks for taking an interest! As you noted, at the end of the video I made a reference to not creating a new buffer unnecessarily - and honestly I still think this is the way to go. Essentially we're asking the computer to do a lot of work that it doesn't need to do.

As far as rendering in react goes, we're always passing immutable references anyway, since the `HexView` slices up a small portion of the buffer. Reacts guidelines about mutation and `setState` are really there to make sure the connection between state and re-render is clear, but in this application we can achieve the same effect in a more performant way since whenever the buffer changes through mutation, some other integral part of state changes as well, guaranteeing that the connection is made. If we wanted to make this more explicit while retaining the similar performance, using a `ref` to hold the buffer state, and creating a hook like `const refreshState = useStateRefresh()` would probably be the way to go. That way, we can still pass the buffer into `useEffect` dependency lists reliably for when we actually resize the buffer.

And I appreciate that you took the time to set up a benchmark too! That's always a great idea :smile: However a benchmark really needs to compare at least two things in order to be meaningful. Your test takes 112ms, but perhaps using mutation takes 1ms. That would be a pretty significant difference!
Also, 16KiB is not a particularly big binary. For this kind of stress-test, we'd want to assess a possible worst case performance - so trying with different files from 20-100MiB would be a more realistic outlook.

francisrstokes commented:
As an example for performance measurement, I put this code together. 25MiB filesize, 1000 write operations at a random index with a random byte (to try to simulate the random access patterns of editing).

```typescript
import { performance, PerformanceObserver } from 'perf_hooks';

const KiB = 1024;
const MiB = 1024 * KiB;

const NUM_ITERATIONS = 1000;
const BUFFER_SIZE = 25 * MiB;

const randomIndex = () => (Math.random() * BUFFER_SIZE) | 0;
const randomByte = () => (Math.random() * 0x100) | 0;

const perfObserver = new PerformanceObserver((items) => {
  items.getEntries().forEach((entry) => {
    console.log(entry)
  });
});

perfObserver.observe({ entryTypes: ['measure'] });

const initialBuffer = new Uint8Array(Array.from({length: BUFFER_SIZE}, randomByte));
let copyBuffer = initialBuffer.slice();

performance.mark('copyStart');
for (let i = 0; i < NUM_ITERATIONS; i++) {
  copyBuffer = copyBuffer.slice();
  copyBuffer[randomIndex()] = randomByte();
}
performance.mark('copyEnd');

performance.mark('noCopyStart');
for (let i = 0; i < NUM_ITERATIONS; i++) {
  initialBuffer[randomIndex()] = randomByte();
}
performance.mark('noCopyEnd');

performance.measure('copy', 'copyStart', 'copyEnd');
performance.measure('noCopy', 'noCopyStart', 'noCopyEnd');
```

The results (i7-11370H @ 3.30GHz):

```
PerformanceMeasure {
  name: 'copy',
  entryType: 'measure',
  startTime: 3322.431111037731,
  duration: 4820.79607796669,
  detail: null
}
PerformanceMeasure {
  name: 'noCopy',
  entryType: 'measure',
  startTime: 8143.4279890060425,
  duration: 0.1619030237197876,
  detail: null
}
```

So we're talking about 5 orders of magnitude difference here. If we double the file size to 50MiB, it scales non-linearly to 6 orders of magnitude.

As you can see, it's definitely worth not creating copies of the buffer if we don't need to!

AhmadMayo commented:
It's now clear to me that this pull request is more of a personal preference than an actual bug fix. Thanks for the detailed response.

francisrstokes commented:
In any case, thanks for opening the issue. Like I say, I think that the observation that mutation is being used in combination with `setState` is a good one - something that would perhaps be hard to see the reasoning for when reading the code - and using something like a `ref` to hold and modify the buffer could be a good way of bridging the performance aspects while making the intentions clear in the react context. I would certainly be open to solving the problem that way.

AhmadMayo commented:
I actually wanted to expand more on why I agree with everything you said, but I still think using `setBuffer` is still a better idea, but I didn't want to assume your interest in hearing my opinion 😄.

Well, here goes. First, I always rely on "the system" or "the standards" if that makes sense. Doing this doesn't leave room for different interpretations, or relying on memory to remember why it was done this way, or to remember that if any one is going to make an update to this part - including future me - they should remember to do it this specific way.

As for the performance, people don't make hundreds of actions per second. So scaling down the timing to normal typing speed - 40 wpm, so I think 5 edits per second is a reasonable guess - it would take around...20ms I guess? Still insignificant.

But again, I have my own experience - good and bad - which shaped my way of thinking, and this effects which problems I think are more important than others, and it's ok if other people don't share my POV.

I'll make another pull request where I'll use a ref instead of a state. May be even abstract it in a new hook?

Commit:
use ref instead of state for performance optimization

AhmadMayo commented:
Done. Check the `use-buffer` file and tell me what you think.

francisrstokes commented:
> I didn't want to assume your interest in hearing my opinion 😄.

I'm definitely interested! If my response came across too blunt then I apologise for that. I'm always happy if someone takes an interest in contributing to the project - perhaps I dive into the details of the conversation too quickly 😄 

> First, I always rely on "the system" or "the standards" if that makes sense. Doing this doesn't leave room for different interpretations, or relying on memory to remember why it was done this way, or to remember that if any one is going to make an update to this part - including future me - they should remember to do it this specific way.

Absolutely - this is also my default position.

> As for the performance, people don't make hundreds of actions per second

True. My default position is also that in most cases, performance isn't a primary concern. In the context of an editor however, I do think it's important to keep the interface as responsive and interactive as possible. You can imagine that when editing a large binary file, you might want to zero some block of memory. We could easily end up allocating/de-allocating 100s of buffers x 100s of MiB. This will definitely have an impact on the user in a single threaded environment - even something like 150ms delay is visible and irritating to a user. It could also create a memory strain if a lot of reallocations are executed in a short time, and old buffers are not garbage collected quickly enough.

But I actually think we can get the best of all worlds in the end 😁 We get transparency in the code by doing what we've discussed with the ref, and outside of this module we expose a higher level interface (insert, update, delete etc).

> But again, I have my own experience - good and bad - which shaped my way of thinking, and this effects which problems I think are more important than others, and it's ok if other people don't share my POV.

I think this is a great attitude to have and I respect it. Fundamentally building software is about being good at assessing and making trade-offs, and this can only really be based on experience. Any given projects requirements will end up giving weight to which way the trade off should go, and in `bewitched`s case, I really think that having a fast core data structure is important for the tool to be fulfilling it's goal. But that was never stated anywhere explicitly, so I don't think it's strange that you would come with questions!

francisrstokes commented on src/hooks/use-edit.ts:
```diff
@@ -24,24 +25,27 @@ export const useEdit = ({
   useInput((input, key) => {
     if (isHexChar(input)) {
       const value = parseInt(input, 16);
+      const newBuffer = buffer.slice();
```
We shouldn't need to do this anymore 🙂

francisrstokes commented:
I left a couple of comments. If you think it makes sense, in `useBuffer` we could expose an `updateAtCursor` function, which can call `setBuffer` internally - then we would only export a high level interface to the rest of the application.

Commit:
expose a high level `updateAtCursor` function instead of `setBuffer`

Commit:
whitespaces

AhmadMayo commented:
> I'm definitely interested! If my response came across too blunt then I apologise for that. I'm always happy if someone takes an interest in contributing to the project - perhaps I dive into the details of the conversation too quickly 😄

On the contrary. I assumed the worst, but your responses were most welcoming.

Check again now. I added `buffer` to the list of the `useEffect`'s dependencies, removed the `setIsMSN(true)` after each `bufferCommand` call, and added an `updateAtCursor` function.

francisrstokes commented:
Awesome, thanks @AhmedSayed77. I've published version 0.3.1 :tada:
